implementations/rust/ockam/ockam_api/src/authenticator/direct.rs

```diff
@@ -0,0 +1,419 @@
+                ["enroller"] => Response::unauthorized(req.id()).to_vec()?,
+                // Enroller wants to add a member.
+                ["member"] => {
+                    if let Some(err) = check_enroller(&self.e_store, &req, from.key_id()).await? {
+                        err.to_vec()?
+                    } else {
+                        let add: AddMember = dec.decode()?;
+                        let member = minicbor::to_vec(Placeholder)?;
+                        self.m_store
+                            .set(MEMBER, add.member().as_str().to_string(), member)
+                            .await?;
+                        Response::ok(req.id()).to_vec()?
+                    }
+                }
+                // Member wants a credential.
+                ["credential"] => {
+                    if let Some(err) = check_member(&self.m_store, &req, from.key_id()).await? {
+                        err.to_vec()?
+                    } else {
+                        let mut attrs = Attributes::new();
+                        attrs.put("id", from.key_id().as_bytes());
+                        let ts = Timestamp::now().ok_or_else(invalid_sys_time)?;
+                        let mut timestamp = Cursor::new([0; 9]);
+                        minicbor::encode(ts, &mut timestamp)?;
+                        attrs.put("ts", &timestamp.get_ref()[..timestamp.position()]);
+                        let crd = self.signer.sign(&attrs).await?;
+                        Response::ok(req.id()).body(crd).to_vec()?
+                    }
+                }
+                _ => response::unknown_path(&req).to_vec()?,
+            },
+            Some(Method::Get) => match req.path_segments::<3>().as_slice() {
+                // Admin wants to check enroller data.
+                ["enroller", id] if self.is_admin(from) => {
+                    if let Some(data) = self.e_store.get(ENROLLER, id).await? {
+                        let enroller = minicbor::decode::<Placeholder>(&data)?;
+                        Response::ok(req.id()).body(enroller).to_vec()?
+                    } else {
+                        Response::not_found(req.id()).to_vec()?
+                    }
+                }
+                ["enroller", _] => Response::unauthorized(req.id()).to_vec()?,
+                // Enroller checks member data.
+                ["member", id] => {
+                    if let Some(err) = check_enroller(&self.e_store, &req, from.key_id()).await? {
+                        err.to_vec()?
+                    } else if let Some(data) = self.m_store.get(MEMBER, id).await? {
+                        let member = minicbor::decode::<Placeholder>(&data)?;
+                        Response::ok(req.id()).body(member).to_vec()?
+                    } else {
+                        Response::not_found(req.id()).to_vec()?
+                    }
+                }
+                // Validate member credential.
+                ["credential"] => {
```

Comment:
Just noting: there must exist check not only if credential is valid, but also if it was issued by known authority and if it was issued to the party presenting this credential. The latter check may be performed here as well